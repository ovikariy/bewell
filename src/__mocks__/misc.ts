/*
  special mock for react-native-get-random-values otherwise gives:
  TypeError: Cannot read property 'getRandomBase64' of undefined
  at Object.<anonymous> (node_modules/react-native-get-random-values/getRandomBase64.js:1:18)
 */
jest.mock('react-native-get-random-values', () => ({
  getRandomBase64: jest.fn(),
}));

/* always keep date now the same because it would change in the test each time it runs and fail */
Date.now = jest.fn(() => 1482363367071);

export const StableDate = new Date(Date.now());