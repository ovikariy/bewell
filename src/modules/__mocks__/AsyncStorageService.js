const testData = {
    // 'key1': 'value1',
    // 'key2': 'value2'
};

export default function mockAsyncStorageService() {
    return {
        getItem: jest.fn((key) => { return testData[key]; }),
        multiGet: jest.fn((keys) => {
            const result = [];
            for(var i=0; i < keys.length; i++) {
                result.push([keys[i], testData[keys[i]]]);
            }
            return result;
        }),
        multiSet: jest.fn((items) => {
            items.forEach((item) => {
                testData[item[0]] = item[1];
            });
        }),
        multiRemove: jest.fn((keys) => {
            keys.forEach((key) => {
                delete testData[key];
            })
        }),
        getAllKeys: jest.fn(() => { return Object.keys(testData); })
    };
}