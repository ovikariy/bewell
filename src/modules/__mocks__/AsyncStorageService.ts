const data: { [key: string]: string } = {};

export function createMockAsyncStorageService() {
    return {
        getItem: jest.fn((key: string) => data[key]),
        multiGet: jest.fn((keys: string[]) => keys.map(key => data[key])),
        multiSet: jest.fn((items: [string, string][]) => {
            items.forEach(item => {
                data[item[0]] = item[1];
            });
        }),
        multiRemove: jest.fn((keys: string[]) => {
            keys.forEach(key => {
                delete data[key];
            });
        }),
        getAllKeys: jest.fn(() => Object.keys(data))
    };
}