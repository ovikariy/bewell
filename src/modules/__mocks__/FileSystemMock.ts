import { isNullOrEmpty } from "../utils";

interface FSInterface {
  freeDiskSpace: number,
  files: { [key: string]: string | null },
  directories: { [key: string]: string | null }
}

export const fs: FSInterface = {
  freeDiskSpace: 999999999999999,
  files: {
    file1: '{ key1: \'value1\' }',
    file2: '{ key2: \'value2\' }',
  },
  directories: {
    cacheDirectory: 'cache/',
    documentDirectory: 'documents/'
  }
};

export function createFileSystemMock() {
  return {
    EncodingType: {
      UTF8: "utf8",
      Base64: "base64"
    },
    WritingOptions: {
      encoding: {
        UTF8: "utf8",
        Base64: "base64"
      }
    },
    cacheDirectory: fs.directories.cacheDirectory,
    documentDirectory: fs.directories.documentDirectory,
    readAsStringAsync: jest.fn((fileUri: string, options?: any) => fs.files[fileUri] ? fs.files[fileUri] + '' : undefined),
    getFreeDiskStorageAsync: jest.fn(() => fs.freeDiskSpace),
    writeAsStringAsync: jest.fn((fileUri: string, contents: string, options?: any) => {
      fs.files[fileUri] = contents;
    }),
    getInfoAsync: jest.fn((fileUri: string, options?: { md5?: boolean, size?: boolean }) => {
      return {
        exists: (fs.files[fileUri] !== undefined || fs.directories[fileUri] !== undefined),
        size: (fs.files[fileUri] !== undefined || fs.directories[fileUri] !== undefined) ? 100 : 0,
        isDirectory: (fs.directories[fileUri] !== undefined) ? true : false,
        uri: fileUri
      };
    }),
    deleteAsync: jest.fn((fileUri: string) => {
      if (fs.files[fileUri])
        delete fs.files[fileUri];
    }),
    makeDirectoryAsync: jest.fn((fileUri: string, options?: { intermediates?: boolean }) => {
      fs.directories[fileUri] = fileUri;
    }),
    readDirectoryAsync: jest.fn((directoryUri: string) => {
      if (isNullOrEmpty(directoryUri))
        return [];
      const result = [];
      if (!directoryUri.endsWith('/') && !directoryUri.endsWith('\\'))
        directoryUri = directoryUri + '/';
      for (const key in fs.files) {
        if (key.startsWith(directoryUri))
          result.push(key.replace(directoryUri, '')); //we want only file names not full paths
      }
      return result;
    }),
    copyAsync: jest.fn((options: { from: string, to: string }) => {
      fs.files[options.to] = fs.files[options.from];
    })
  };
}