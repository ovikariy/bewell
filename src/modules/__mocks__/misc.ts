/*
  special mock for react-native-get-random-values otherwise gives:
  TypeError: Cannot read property 'getRandomBase64' of undefined
  at Object.<anonymous> (node_modules/react-native-get-random-values/getRandomBase64.js:1:18)
 */
jest.mock('react-native-get-random-values', () => ({
  getRandomBase64: jest.fn(),
}));