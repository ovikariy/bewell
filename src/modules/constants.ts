import { ErrorMessage } from "./translations";
export { ErrorMessage };

export const images = {
  logo_light: require('../assets/images/logo-light.png'),
  logo_dark: require('../assets/images/logo-dark.png'),
  logo_small_light: require('../assets/images/logo-small-light.png'),
  logo_small_dark: require('../assets/images/logo-small-dark.png'),
  splash: require('../assets/images/splash.png'),
  icon: require('../assets/images/icon.png'),
  header: getRandomHeaderImage(), /* chooses a random image from the folder for fun */
  arrow_up_dark: require('../assets/images/arrow-up-dark.png'),
  arrow_up_light: require('../assets/images/arrow-up-light.png'),
};

/* ItemTypes correspond to data keys in storage and will be hashed with DATA ENCRYPTION KEY
and their values will be encrypted with the DATA ENCRYPTION KEY;
These are only for widgets and not to be used for other storage items */
export const ItemTypes = {
  MOOD: 'MOOD',
  NOTE: 'NOTE',
  SLEEP: 'SLEEP',
  IMAGE: 'IMAGE'
};

export const brokenImageURI = 'broken'; /* this is just so that we can show something to user to indicate the image will not get loaded */

export const stateConstants = {
  STORE: 'STORE',
  OPERATION: 'OPERATION',
  AUTH: 'AUTH',
  PINSETUP: 'PINSETUP',
  BACKUPRESTORE: 'BACKUPRESTORE',
  CHANGEPASSWORD: 'CHANGEPASSWORD',
  APPCONTEXT: 'APPCONTEXT'
};

export class StoreConstants {
  static isInitialized = 'isInitialized';
  static loginAttempts = 'loginAttempts';
  static maxLoginAttempts = 6;
  /* In storage we use 'bewellapp_key' pattern for keys in key/value pairs */
  /* DataEncryptionStoreKey is special as it will not be hashed
  and the value will be encrypted with the user's password. When the user changes password
  only DataEncryptionStoreKey will have to be re-encrypted in store */
  static keyPrefix = 'bewellapp_';
  static keyDateFormat = 'MMYYYY';
  static monthsFromEpochDate = getMonthsFromEpochDate(StoreConstants.keyPrefix);
  static AllEncryptedStoreKeys = [...StoreConstants.monthsFromEpochDate];
  static DataEncryptionStoreKey = StoreConstants.keyPrefix + 'DATAENCRYPTIONKEY';
  static SETTINGS = StoreConstants.keyPrefix + 'SETTINGS';
}

function getMonthsFromEpochDate(keyPrefix: string) {
  /* data in store is partitioned by month so the keys can be potentially from
  when the app was first used to current year and month  ['012019', '022019', ... ] */
  const epochYear = 2019;
  const currentYear = new Date().getUTCFullYear();
  const currentMonth = new Date().getUTCMonth() + 1; /* getUTCMonth is zero based */
  const result = [];

  for (let year = epochYear; year <= currentYear; year++) {
    for (let month = 1; month <= 12; month++) {
      if (year === currentYear && month > currentMonth)
        break;
      result.push(keyPrefix + (month < 10 ? '0' : '') + month + '' + year); // 'MMYYYY' format
    };
  }
  return result;
}

export const settingsConstants = {
  language: 'language',
  theme: 'theme',
  version: 'version',
  hideNoteText: 'hideNoteText'
};

export enum ErrorCode {
  Encrypt1 = 'E1001',
  Encrypt2 = 'E1002',
  Encrypt3 = 'E1003',
  Encrypt4 = 'E1004',
  Decrypt1 = 'D1001',
  Decrypt2 = 'D1002',
  Decrypt3 = 'D1003',
  Decrypt4 = 'D1004',
  Decrypt5 = 'D1005',
  Decrypt6 = 'D1006',
  Decrypt7 = 'D1007',
  Decrypt8 = 'D1008',
  Decrypt9 = 'D1009',
  Decrypt10 = 'D10010',
  MissingItemType1 = 'MIT1001',
  MissingKey1 = 'MK1001',
  MissingKey2 = 'MK1002',
  MissingKey3 = 'MK1003',
  MissingKey4 = 'MK1004',
  MissingKey5 = 'MK1005',
  MissingKey7 = 'MK1007',
  MissingKey8 = 'MK1008',
  MissingKey9 = 'MK1009',
  MissingKey10 = 'MK10010',
  MissingKey11 = 'MK10011',
  UnableToHashWithoutPassword = 'HH1001',
  Hash1 = 'H1001',
  Hash2 = 'H1002',
  Storage1 = 'S1001',
  Storage2 = 'S1002',
  Storage3 = 'S1003',
  Storage4 = 'S1004',
  Storage5 = 'S1005',
  Storage6 = 'S1006',
  Storage7 = 'S1007',
  Storage8 = 'S108',
  Storage9 = 'S109',
  Storage10 = 'S1010',
  Storage11 = 'S1011',
  Storage12 = 'S1012',
  Auth1 = 'A1001',
  Auth2 = 'A1002',
  Auth3 = 'A1003',
  Auth4 = 'A1004',
  Auth5 = 'A1005',
  Auth6 = 'A1006',
  Auth7 = 'A1007',
  Auth8 = 'A1008',
  Auth9 = 'A1009',
  Auth10 = 'A10010',
  Auth11 = 'A10011',
  Auth12 = 'A10012',
  Security1 = 'SE1001',
  Security2 = 'SE1002',
  Security3 = 'SE1003',
  Security4 = 'SE1004',
  Security5 = 'SE1005',
  Security6 = 'SE1006',
  Security7 = 'SE1007',
  Security8 = 'SE1008',
  Security9 = 'SE1009',
  Security10 = 'SE1010',
  Import1 = 'I1001',
  Import2 = 'I1002',
  Import3 = 'I1003',
  BackupRestore1 = 'BR1001',
  BackupRestore2 = 'BR1002',
  BackupRestore3 = 'BR1003',
  BackupRestore4 = 'BR1004',
  File1 = 'F1001',
  File2 = 'F1002',
  File3 = 'F1003',
  File4 = 'F1004'
};
/**
 * @description Chooses an image from assets/images/headers directory
 * based on a random number and image filename. This is just for fun
 * instead of always showing the same header image. The filenames should
 * be 'header1.png', 'header2.png' etc format
 */
function getRandomHeaderImage() {
  const randomNumberForHeader = Math.floor(Math.random() * 11);  // returns a random integer from 0 to 10
  let headerImage;
  switch (randomNumberForHeader) {
    case 0:
      headerImage = require('../assets/images/headers/header1.png');
      break;
    case 1:
      headerImage = require('../assets/images/headers/header1.png');
      break;
    case 2:
      headerImage = require('../assets/images/headers/header2.png');
      break;
    case 3:
      headerImage = require('../assets/images/headers/header3.png');
      break;
    case 4:
      headerImage = require('../assets/images/headers/header4.png');
      break;
    case 5:
      headerImage = require('../assets/images/headers/header5.png');
      break;
    case 6:
      headerImage = require('../assets/images/headers/header6.png');
      break;
    case 7:
      headerImage = require('../assets/images/headers/header7.png');
      break;
    case 8:
      headerImage = require('../assets/images/headers/header8.png');
      break;
    case 9:
      headerImage = require('../assets/images/headers/header9.png');
      break;
    case 10:
      headerImage = require('../assets/images/headers/header10.png');
      break;
    default:
      headerImage = require('../assets/images/headers/header1.png');
      break;
  }
  return headerImage;
}