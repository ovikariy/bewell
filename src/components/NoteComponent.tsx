import React, { Component } from 'react';
import { ClearTextArea, IconForButton, TimePicker } from './MiscComponents';
import { AppContext } from '../modules/AppContext';
import { WidgetBase, WidgetComponentPropsBase } from '../modules/WidgetFactory';
import { Text, View } from 'react-native';
import { friendlyTime } from '../modules/helpers';

export interface NoteComponentWidgetType extends WidgetBase {
  note?: string
}

export interface NoteComponentProps extends WidgetComponentPropsBase {
  value: NoteComponentWidgetType;
  onChange: (newValue: NoteComponentWidgetType) => void
}

export const NoteComponent = (props: NoteComponentProps) => {
  const context = React.useContext(AppContext);
  const language = context.language;
  return <ClearTextArea
    numberOfLines={1}
    placeholder={language.whatsOnYourMind}
    value={props.value ? props.value.note : undefined}
    onChangeText={(note) => { props.onChange({ ...props.value, note }) }}
  />
}

interface NoteHistoryComponentProps {
  item: NoteComponentWidgetType;
  isSelectedItem: boolean;
}
export const NoteHistoryComponent = (props: NoteHistoryComponentProps) => {
  const context = React.useContext(AppContext);
  const styles = context.styles;

  return (<View style={styles.row}>
    <View style={[styles.flex]}>
      <Text style={props.isSelectedItem ? [styles.bodyText, styles.highlightColor] : styles.bodyText}>
        {friendlyTime(props.item.date)}</Text>
      <Text style={props.isSelectedItem ? [styles.subTitleText, styles.highlightColor] : styles.subTitleText}>
        {props.item.note + 'hello'}</Text>
    </View>
  </View>)
}