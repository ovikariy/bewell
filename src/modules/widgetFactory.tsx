import React, { ReactNode } from 'react';
import { ItemTypes, settingsConstants, settingsLists } from './constants';
import { AppContextState } from '../redux/reducerTypes';
import { NoteComponent, NoteComponentProps, NoteComponentWidgetType, NoteHistoryComponent } from '../components/NoteComponent';
import { CustomIconRatingCalendarComponent, CustomIconRatingComponent, CustomIconRatingComponentProps, CustomIconRatingComponentWidgetType, CustomIconRatingHistoryComponent } from '../components/CustomIconRatingComponent';
import { SleepComponent, SleepComponentProps, SleepComponentWidgetType, SleepHistoryComponent, SleepCalendarComponent } from '../components/SleepComponent';
import { ImagePickerComponent, ImagePickerComponentProps, ImagePickerWidgetType, ImagePickerHistoryComponent } from '../components/ImagePickerComponent';
import { StyleProp, ViewStyle } from 'react-native';
import { CustomIconType } from '../components/CustomIconRating';
import { WidgetBase } from './types';
import { sizes } from '../assets/styles/style';
import { RatingCalendarComponent, RatingComponent, RatingComponentProps, RatingComponentWidgetType, RatingHistoryComponent } from '../components/RatingComponent';
import { ActivityCalendarComponent, ActivityComponent, ActivityComponentProps, ActivityComponentWidgetType, ActivityHistoryComponent } from '../components/ActivityComponent';

export function CreateWidgetFactory(context: AppContextState) {
  const language = context.language;
  const styles = context.styles;

  const widgetFactory: WidgetFactory = {
    [ItemTypes.NOTE]:
      {
        config: {
          widgetTitle: language.note,
          historyTitle: language.notes,
          itemTypeName: ItemTypes.NOTE,
          addIcon: { text: language.note, name: 'pencil-square-o', type: 'font-awesome' } as WidgetAddIconConfig,
          style: {} as ViewStyle
        } as WidgetConfig,
        renderWidgetItem: (props: NoteComponentProps, config: WidgetConfig) => {
          return <NoteComponent  {...props} config={config} />;
        },
        renderHistoryItem: (item: NoteComponentWidgetType, isSelectedItem: boolean, config: WidgetConfig) => {
          return <NoteHistoryComponent item={item} isSelectedItem={isSelectedItem} />;
        }
      } as WidgetFactoryType,
    [ItemTypes.MOOD]:
      {
        config: {
          widgetTitle: language.mood,
          historyTitle: language.moods,
          itemTypeName: ItemTypes.MOOD,
          isHorizontalHistoryRow: true, /* on the history screen, show all items for the day in one row */
          addIcon: { text: language.mood, name: 'smile-o', type: 'font-awesome' } as WidgetAddIconConfig,
          style: {} as ViewStyle,
          icons: [
            { name: language.happy, icon: 'mood-happy', iconStyle: styles.brightColor, backgroundStyle: { backgroundColor: '#ff9a55' } } as CustomIconType,
            { name: language.soso, icon: 'mood-neutral', iconStyle: styles.brightColor, backgroundStyle: { backgroundColor: '#009898' } } as CustomIconType,
            { name: language.couldBeBetter, icon: 'mood-sad', iconStyle: styles.brightColor, backgroundStyle: { backgroundColor: '#517fa4' } } as CustomIconType
          ]
        },
        renderWidgetItem: (props: CustomIconRatingComponentProps, config: WidgetConfig) => {
          return <CustomIconRatingComponent  {...props} config={config} />;
        },
        renderHistoryItem: (item: CustomIconRatingComponentWidgetType, isSelectedItem: boolean, config: WidgetConfig) => {
          return <CustomIconRatingHistoryComponent item={item} isSelectedItem={isSelectedItem} config={config} />;
        },
        renderCalendarItem: (item: CustomIconRatingComponentWidgetType, config: WidgetConfig) => {
          return <CustomIconRatingCalendarComponent item={item} isSelectedItem={false} config={config} />;
        }
      } as WidgetFactoryType,
    [ItemTypes.SLEEP]:
      {
        config: {
          widgetTitle: language.sleep,
          historyTitle: language.sleeps,
          itemTypeName: ItemTypes.SLEEP,
          addIcon: { text: language.sleep, name: 'moon-o', type: 'font-awesome' } as WidgetAddIconConfig,
          style: {} as ViewStyle,
          icons: [
            { name: language.restful, icon: 'sleep-happy', iconStyle: styles.brightColor, backgroundStyle: { backgroundColor: '#ff9a55' } } as CustomIconType,
            { name: language.interrupted, icon: 'sleep-neutral', iconStyle: styles.brightColor, backgroundStyle: { backgroundColor: '#009898' } } as CustomIconType,
            { name: language.poor, icon: 'sleep-sad', iconStyle: styles.brightColor, backgroundStyle: { backgroundColor: '#517fa4' } } as CustomIconType
          ]
        } as WidgetConfig,
        renderWidgetItem: (props: SleepComponentProps, config: WidgetConfig) => {
          return <SleepComponent {...props} config={config} />;
        },
        renderHistoryItem: (item: SleepComponentWidgetType, isSelectedItem: boolean, config: WidgetConfig) => {
          return <SleepHistoryComponent item={item} isSelectedItem={isSelectedItem} config={config} />;
        },
        renderCalendarItem: (item: SleepComponentWidgetType, config: WidgetConfig) => {
          return <SleepCalendarComponent item={item} isSelectedItem={false} config={config} />;
        }
      } as WidgetFactoryType,
    [ItemTypes.IMAGE]:
      {
        config: {
          widgetTitle: language.image,
          hideTitleInHeader: true,
          historyTitle: language.images,
          itemTypeName: ItemTypes.IMAGE,
          addIcon: { text: language.image, name: 'picture-o', type: 'font-awesome' } as WidgetAddIconConfig,
          style: { paddingVertical: 0, paddingTop: sizes[20] } as ViewStyle
        } as WidgetConfig,
        renderWidgetItem: (props: ImagePickerComponentProps, config: WidgetConfig) => {
          return <ImagePickerComponent  {...props} config={config} readonly={false} />;
        },
        renderHistoryItem: (item: ImagePickerWidgetType, isSelectedItem: boolean, config: WidgetConfig) => {
          return <ImagePickerHistoryComponent item={item} isSelectedItem={isSelectedItem} config={config} />;
        }
      } as WidgetFactoryType,
    [ItemTypes.MOVE]:
      {
        config: {
          widgetTitle: language.movement,
          historyTitle: language.movement,
          itemTypeName: ItemTypes.MOVE,
          addIcon: { text: language.move, name: 'walk', type: 'material-community' } as WidgetAddIconConfig,
          style: {} as ViewStyle
        },
        renderWidgetItem: (props: ActivityComponentProps, config: WidgetConfig) => {
          return <ActivityComponent activities={settingsLists.exercises} showDistance={true} showDuration={true} {...props} config={config} />;
        },
        renderHistoryItem: (item: ActivityComponentWidgetType, isSelectedItem: boolean, config: WidgetConfig) => {
          return <ActivityHistoryComponent item={item} isSelectedItem={isSelectedItem} config={config} />;
        },
        renderCalendarItem: (item: ActivityComponentWidgetType, config: WidgetConfig) => {
          return <ActivityCalendarComponent item={item} isSelectedItem={false} config={config} />;
        }
      } as WidgetFactoryType,
    [ItemTypes.WATER]:
      {
        config: {
          widgetTitle: language.waterIntake,
          historyTitle: language.waterIntake,
          itemTypeName: ItemTypes.WATER,
          addIcon: { text: language.water, name: 'cup', type: 'material-community' } as WidgetAddIconConfig,
          style: {} as ViewStyle,
          ratings: [1, 2, 3, 4, 5, 6, 7, 8]
        },
        renderWidgetItem: (props: RatingComponentProps, config: WidgetConfig) => {
          return <RatingComponent iconName='cup' {...props} config={config} />;
        },
        renderHistoryItem: (item: RatingComponentWidgetType, isSelectedItem: boolean, config: WidgetConfig) => {
          return <RatingHistoryComponent iconName='cup' item={item} isSelectedItem={isSelectedItem} config={config} />;
        },
        renderCalendarItem: (item: RatingComponentWidgetType, config: WidgetConfig) => {
          return <RatingCalendarComponent iconName='cup' item={item} isSelectedItem={false} config={config} />;
        }
      } as WidgetFactoryType,
    [ItemTypes.CREATE]:
      {
        config: {
          widgetTitle: language.creativity,
          historyTitle: language.creativity,
          itemTypeName: ItemTypes.CREATE,
          addIcon: { text: language.create, name: 'creation', type: 'material-community' } as WidgetAddIconConfig,
          style: {} as ViewStyle
        },
        renderWidgetItem: (props: ActivityComponentProps, config: WidgetConfig) => {
          return <ActivityComponent activities={settingsLists.creativity} showDuration={true} {...props} config={config} />;
        },
        renderHistoryItem: (item: ActivityComponentWidgetType, isSelectedItem: boolean, config: WidgetConfig) => {
          return <ActivityHistoryComponent item={item} isSelectedItem={isSelectedItem} config={config} />;
        },
        renderCalendarItem: (item: ActivityComponentWidgetType, config: WidgetConfig) => {
          return <ActivityCalendarComponent item={item} isSelectedItem={false} config={config} />;
        }
      } as WidgetFactoryType,
    [ItemTypes.MEDITATE]:
      {
        config: {
          widgetTitle: language.meditationMinutes,
          historyTitle: language.meditation,
          itemTypeName: ItemTypes.MEDITATE,
          addIcon: { text: language.meditate, name: 'spa', type: 'material-community' } as WidgetAddIconConfig,
          style: {} as ViewStyle,
          ratings: [1, 3, 5, 10, 15, 20, 30, 60]
        } as WidgetConfig,
        renderWidgetItem: (props: RatingComponentProps, config: WidgetConfig & { ratings: number[] }) => {
          return <RatingComponent iconName='spa' {...props} config={config} />;
        },
        renderHistoryItem: (item: RatingComponentWidgetType, isSelectedItem: boolean, config: WidgetConfig) => {
          return <RatingHistoryComponent iconName='spa' item={item} isSelectedItem={isSelectedItem} config={config} />;
        },
        renderCalendarItem: (item: RatingComponentWidgetType, config: WidgetConfig) => {
          return <RatingCalendarComponent iconName='spa' item={item} isSelectedItem={false} config={config} hideText={true} />;
        }
      } as WidgetFactoryType,
    [ItemTypes.PERIOD]:
      {
        config: {
          widgetTitle: language.period,
          historyTitle:  language.period,
          itemTypeName: ItemTypes.PERIOD,
          isHorizontalHistoryRow: true, /* on the history screen, show all items for the day in one row */
          addIcon: { text: language.period, name: 'water', type: 'material-community' } as WidgetAddIconConfig,
          style: {} as ViewStyle,
          icons: [
            { name: 'Light', icon: 'period-light', iconStyle: styles.brightColor, backgroundStyle: { backgroundColor: '#FCB99B' } } as CustomIconType,
            { name: 'Medium', icon: 'period-medium', iconStyle: styles.brightColor, backgroundStyle: { backgroundColor: '#FE9D72' } } as CustomIconType,
            { name: 'Heavy', icon: 'period-heavy', iconStyle: styles.brightColor, backgroundStyle: { backgroundColor: '#F67E49' } } as CustomIconType
          ]
        },
        renderWidgetItem: (props: CustomIconRatingComponentProps, config: WidgetConfig) => {
          return <CustomIconRatingComponent  {...props} config={config} />;
        },
        renderHistoryItem: (item: CustomIconRatingComponentWidgetType, isSelectedItem: boolean, config: WidgetConfig) => {
          return <CustomIconRatingHistoryComponent item={item} isSelectedItem={isSelectedItem} config={config} />;
        },
        renderCalendarItem: (item: CustomIconRatingComponentWidgetType, config: WidgetConfig) => {
          return <CustomIconRatingCalendarComponent item={item} isSelectedItem={false} config={config} />;
        }
      } as WidgetFactoryType,
  };

  return widgetFactory;
}

export interface WidgetComponentPropsBase {
  value: WidgetBase;
  selectedDate: Date,
  isSelected?: boolean;
  config: WidgetConfig;
  onChange?: (newValue: WidgetBase) => void
  onSelected?: (id: string) => void;
}

/** Config for icon that shows up in widget toolbar to add a new widget to the list of widgets */
export interface WidgetAddIconConfig {
  text: string;
  name: string;
  type: string;
}

/** General configuration for widgets e.g. title used on the history screen, styles etc */
export interface WidgetConfig {
  widgetTitle: string;
  historyTitle: string;
  itemTypeName: string;
  addIcon: WidgetAddIconConfig;
  style?: StyleProp<ViewStyle>;
  hideTitleInHeader?: boolean;
  isHorizontalHistoryRow?: boolean, /* on the history screen, show all items for the day in one row */
  icons?: CustomIconType[]
}

export interface WidgetFactory {
  [key: string]: WidgetFactoryType
}

export interface WidgetFactoryType {
  config: WidgetConfig;
  renderWidgetItem: (props: WidgetComponentPropsBase, config: WidgetConfig) => ReactNode;
  renderHistoryItem?: (item: WidgetBase, isSelectedItem: boolean, config: WidgetConfig) => ReactNode;
  renderCalendarItem?: (item: WidgetBase, config: WidgetConfig) => ReactNode;
}