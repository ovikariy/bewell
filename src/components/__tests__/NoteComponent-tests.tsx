import 'react-native';
import * as mocks from '../../__mocks__/misc'; /** special mocks */
import React from 'react';
import renderer from 'react-test-renderer';
import { NoteComponent, NoteComponentWidgetType } from '../NoteComponent';
import { ItemTypes } from '../../modules/constants';
import { CreateWidgetFactory } from '../../modules/widgetFactory';
import { defaultAppContext } from '../../modules/appContext';

it('NoteComponent renders correctly', () => {
  const note: NoteComponentWidgetType = {
    date: mocks.StableDate.toUTCString(),
    dateCreated: mocks.StableDate.toUTCString(),
    id: 'sleep1',
    type: ItemTypes.NOTE
  };
  const widgetFactory = CreateWidgetFactory(defaultAppContext);

  expect(renderer.create(
    <NoteComponent value={note} onChange={() => { }} config={widgetFactory[ItemTypes.NOTE].config} selectedDate={mocks.StableDate} />
  )).toMatchSnapshot();

  note.note = `"testing note's line 1
  line 2
  line 3
  "`;

  expect(renderer.create(
    <NoteComponent value={note} onChange={() => { }} config={widgetFactory[ItemTypes.NOTE].config} selectedDate={mocks.StableDate}
      isSelected={true} key={note.id} onSelected={() => { }} />
  )).toMatchSnapshot();
});
