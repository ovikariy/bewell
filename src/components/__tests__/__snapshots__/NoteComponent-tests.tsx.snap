// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteComponent renders correctly 1`] = `
<View
  style={
    Object {
      "paddingHorizontal": 10,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomWidth": 0,
        "borderColor": "#86939e",
        "borderWidth": 0,
        "flexDirection": "row",
        "margin": 0,
        "padding": 0,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoFocus={true}
      editable={true}
      multiline={true}
      numberOfLines={1}
      onChangeText={[Function]}
      placeholder="What's on your mind?"
      placeholderTextColor="#FFFFFF70"
      rejectResponderTermination={true}
      scrollEnabled={false}
      style={
        Object {
          "alignSelf": "center",
          "color": "#87BCBF",
          "flex": 1,
          "fontFamily": "Helvetica",
          "fontSize": 17,
          "minHeight": 40,
        }
      }
      testID="RNE__Input__text-input"
      textAlignVertical="top"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;

exports[`NoteComponent renders correctly 2`] = `
<View
  style={
    Object {
      "paddingHorizontal": 10,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomWidth": 0,
        "borderColor": "#86939e",
        "borderWidth": 0,
        "flexDirection": "row",
        "margin": 0,
        "padding": 0,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoFocus={false}
      editable={true}
      multiline={true}
      numberOfLines={1}
      onChangeText={[Function]}
      placeholder="What's on your mind?"
      placeholderTextColor="#FFFFFF70"
      rejectResponderTermination={true}
      scrollEnabled={false}
      style={
        Object {
          "alignSelf": "center",
          "color": "#87BCBF",
          "flex": 1,
          "fontFamily": "Helvetica",
          "fontSize": 17,
          "minHeight": 40,
        }
      }
      testID="RNE__Input__text-input"
      textAlignVertical="top"
      underlineColorAndroid="transparent"
      value="\\"testing note's line 1
  line 2
  line 3
  \\""
    />
  </View>
</View>
`;
